File: android/app/src/main/java/com/purchasely/shaker/ui/screen/settings/SettingsViewModel.kt
package com.purchasely.shaker.ui.screen.settings

import android.app.Activity
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.purchasely.shaker.data.PurchaselySdkMode
import com.purchasely.shaker.domain.model.DisplayMode
import com.purchasely.shaker.domain.model.ThemeMode
import com.purchasely.shaker.domain.repository.PremiumRepository
import com.purchasely.shaker.data.RunningModeRepository
import com.purchasely.shaker.data.SettingsRepository
import com.purchasely.shaker.purchasely.DisplayResult
import com.purchasely.shaker.purchasely.FetchResult
import com.purchasely.shaker.purchasely.PresentationHandle
import com.purchasely.shaker.purchasely.PurchaselyWrapper
import io.purchasely.ext.PLYDataProcessingPurpose
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class SettingsViewModel(
    private val settingsRepo: SettingsRepository,
    private val premiumRepository: PremiumRepository,
    private val runningModeRepo: RunningModeRepository,
    private val purchaselyWrapper: PurchaselyWrapper
) : ViewModel() {

    private val _userId = MutableStateFlow(settingsRepo.userId)
    val userId: StateFlow<String?> = _userId.asStateFlow()

    val isPremium: StateFlow<Boolean> = premiumRepository.isPremium

    private val _restoreMessage = MutableStateFlow<String?>(null)
    val restoreMessage: StateFlow<String?> = _restoreMessage.asStateFlow()

    private val _themeMode = MutableStateFlow(settingsRepo.themeMode)
    val themeMode: StateFlow<ThemeMode> = _themeMode.asStateFlow()

    private val _sdkMode = MutableStateFlow(
        PurchaselySdkMode.fromStorage(settingsRepo.sdkModeStorage)
    )
    val sdkMode: StateFlow<PurchaselySdkMode> = _sdkMode.asStateFlow()

    // Data privacy consent toggles (default: true = consent given)
    private val _analyticsConsent = MutableStateFlow(settingsRepo.analyticsConsent)
    val analyticsConsent: StateFlow<Boolean> = _analyticsConsent.asStateFlow()

    private val _identifiedAnalyticsConsent = MutableStateFlow(settingsRepo.identifiedAnalyticsConsent)
    val identifiedAnalyticsConsent: StateFlow<Boolean> = _identifiedAnalyticsConsent.asStateFlow()

    private val _personalizationConsent = MutableStateFlow(settingsRepo.personalizationConsent)
    val personalizationConsent: StateFlow<Boolean> = _personalizationConsent.asStateFlow()

    private val _campaignsConsent = MutableStateFlow(settingsRepo.campaignsConsent)
    val campaignsConsent: StateFlow<Boolean> = _campaignsConsent.asStateFlow()

    private val _thirdPartyConsent = MutableStateFlow(settingsRepo.thirdPartyConsent)
    val thirdPartyConsent: StateFlow<Boolean> = _thirdPartyConsent.asStateFlow()

    private val _runningMode = MutableStateFlow(
        if (runningModeRepo.isObserverMode) "observer" else "full"
    )
    val runningMode: StateFlow<String> = _runningMode.asStateFlow()

    private val _anonymousId = MutableStateFlow(purchaselyWrapper.anonymousUserId)
    val anonymousId: StateFlow<String> = _anonymousId.asStateFlow()

    private val _displayMode = MutableStateFlow(settingsRepo.displayMode)
    val displayMode: StateFlow<DisplayMode> = _displayMode.asStateFlow()

    // Signal Screen to display onboarding paywall
    private var pendingOnboardingPresentation: PresentationHandle? = null
    private val _requestPaywallDisplay = MutableSharedFlow<Unit>()
    val requestPaywallDisplay: SharedFlow<Unit> = _requestPaywallDisplay.asSharedFlow()

    val sdkVersion: String get() = purchaselyWrapper.sdkVersion

    init {
        settingsRepo.initSdkModeIfNeeded()
        applyConsentPreferences()
    }

    fun refreshAnonymousId() {
        _anonymousId.value = purchaselyWrapper.anonymousUserId
    }

    fun login(userId: String) {
        if (userId.isBlank()) return

        // PURCHASELY: Associate this session with an authenticated user ID
        // The callback's `refresh` flag indicates whether subscriptions should be re-fetched
        // Docs: https://docs.purchasely.com/quick-start/sdk-configuration/user-login
        purchaselyWrapper.userLogin(userId) { refresh ->
            if (refresh) {
                premiumRepository.refreshPremiumStatus()
            }
            Log.d(TAG, "[Shaker] Logged in as: $userId (refresh: $refresh)")
        }

        _userId.value = userId
        settingsRepo.userId = userId

        // PURCHASELY: Store the user ID as a custom attribute for paywall targeting/personalization
        // Docs: https://docs.purchasely.com/advanced-features/user-attributes
        purchaselyWrapper.setUserAttribute("user_id", userId)
    }

    fun logout() {
        // PURCHASELY: Disassociate the current user — clears cached user attributes and subscriptions
        // Docs: https://docs.purchasely.com/quick-start/sdk-configuration/user-login
        purchaselyWrapper.userLogout()
        _userId.value = null
        settingsRepo.userId = null
        premiumRepository.refreshPremiumStatus()
        Log.d(TAG, "[Shaker] Logged out")
    }

    fun restorePurchases() {
        _restoreMessage.value = null
        // PURCHASELY: Restore previously purchased products for the current user
        // Required by App Store / Play Store guidelines; call on explicit user request only
        // Docs: https://docs.purchasely.com/quick-start/sdk-implementation/restore-purchases
        purchaselyWrapper.restoreAllProducts(
            onSuccess = { plan ->
                premiumRepository.refreshPremiumStatus()
                _restoreMessage.value = "Purchases restored successfully!"
                Log.d(TAG, "[Shaker] Restore success: ${plan?.name}")
            },
            onError = { error ->
                _restoreMessage.value = error?.message ?: "No purchases to restore"
                Log.e(TAG, "[Shaker] Restore error: ${error?.message}")
            }
        )
    }

    fun clearRestoreMessage() {
        _restoreMessage.value = null
    }

    fun onPurchaseCompleted() {
        premiumRepository.refreshPremiumStatus()
    }

    fun showOnboardingPaywall() {
        viewModelScope.launch {
            when (val result = purchaselyWrapper.loadPresentation("onboarding")) {
                is FetchResult.Success -> {
                    pendingOnboardingPresentation = result.handle
                    _requestPaywallDisplay.emit(Unit)
                }
                is FetchResult.Client -> {
                    Log.d(TAG, "[Shaker] CLIENT presentation received for onboarding placement — build custom UI here")
                }
                is FetchResult.Deactivated -> {
                    Log.d(TAG, "[Shaker] Onboarding presentation is deactivated")
                }
                is FetchResult.Error -> {
                    Log.d(TAG, "[Shaker] Onboarding presentation not available: ${result.message}")
                }
            }
        }
    }

    suspend fun displayPendingPaywall(activity: Activity) {
        val handle = pendingOnboardingPresentation ?: return
        pendingOnboardingPresentation = null
        val result = purchaselyWrapper.display(handle, activity)
        when (result) {
            is DisplayResult.Purchased, is DisplayResult.Restored -> {
                Log.d(TAG, "[Shaker] Purchased/Restored from onboarding")
                onPurchaseCompleted()
            }
            else -> {}
        }
    }

    fun setDisplayMode(mode: DisplayMode) {
        _displayMode.value = mode
        settingsRepo.displayMode = mode
        Log.d(TAG, "[Shaker] Display mode changed to: ${mode.storageValue}")
    }

    fun setThemeMode(mode: ThemeMode) {
        _themeMode.value = mode
        settingsRepo.themeMode = mode
        // PURCHASELY: Track the user's preferred theme as a custom attribute for audience segmentation
        // Docs: https://docs.purchasely.com/advanced-features/user-attributes
        purchaselyWrapper.setUserAttribute("app_theme", mode.storageValue)
    }

    fun setSdkMode(mode: PurchaselySdkMode) {
        if (_sdkMode.value == mode) return

        _sdkMode.value = mode
        settingsRepo.sdkModeStorage = mode.storageValue
        restartPurchaselySdk(mode)
    }

    fun setAnalyticsConsent(enabled: Boolean) {
        _analyticsConsent.value = enabled
        settingsRepo.analyticsConsent = enabled
        applyConsentPreferences()
    }

    fun setIdentifiedAnalyticsConsent(enabled: Boolean) {
        _identifiedAnalyticsConsent.value = enabled
        settingsRepo.identifiedAnalyticsConsent = enabled
        applyConsentPreferences()
    }

    fun setPersonalizationConsent(enabled: Boolean) {
        _personalizationConsent.value = enabled
        settingsRepo.personalizationConsent = enabled
        applyConsentPreferences()
    }

    fun setCampaignsConsent(enabled: Boolean) {
        _campaignsConsent.value = enabled
        settingsRepo.campaignsConsent = enabled
        applyConsentPreferences()
    }

    fun setThirdPartyConsent(enabled: Boolean) {
        _thirdPartyConsent.value = enabled
        settingsRepo.thirdPartyConsent = enabled
        applyConsentPreferences()
    }

    private fun restartPurchaselySdk(mode: PurchaselySdkMode) {
        purchaselyWrapper.restart()
        Log.d(TAG, "[Shaker] SDK restarted with mode ${mode.storageValue}")
    }

    private fun applyConsentPreferences() {
        val revoked = mutableSetOf<PLYDataProcessingPurpose>()
        if (!_analyticsConsent.value) revoked.add(PLYDataProcessingPurpose.Analytics)
        if (!_identifiedAnalyticsConsent.value) revoked.add(PLYDataProcessingPurpose.IdentifiedAnalytics)
        if (!_personalizationConsent.value) revoked.add(PLYDataProcessingPurpose.Personalization)
        if (!_campaignsConsent.value) revoked.add(PLYDataProcessingPurpose.Campaigns)
        if (!_thirdPartyConsent.value) revoked.add(PLYDataProcessingPurpose.ThirdPartyIntegrations)
        // PURCHASELY: Revoke GDPR data-processing consent for specific purposes
        // Pass the set of revoked purposes; an empty set re-grants all consent
        // Docs: https://docs.purchasely.com/advanced-features/gdpr
        purchaselyWrapper.revokeDataProcessingConsent(revoked)
        Log.d(TAG, "[Shaker] Consent updated — revoked: $revoked")
    }

    companion object {
        private const val TAG = "SettingsViewModel"
    }
}

File: ios/Shaker/Screens/Settings/SettingsViewModel.swift
import Foundation
import Purchasely

class SettingsViewModel: ObservableObject {

    @Published var userId: String?
    @Published var restoreMessage: String?
    @Published var themeMode: String
    @Published var sdkMode: PurchaselySDKMode
    @Published var sdkModeRestartMessage: String?

    // Data privacy consent (default: true = consent given)
    @Published var analyticsConsent: Bool
    @Published var identifiedAnalyticsConsent: Bool
    @Published var personalizationConsent: Bool
    @Published var campaignsConsent: Bool
    @Published var thirdPartyConsent: Bool
    @Published var runningMode: String
    @Published var anonymousId: String = ""
    @Published var displayMode: String

    // Prefetched onboarding presentation
    @Published var onboardingFetchResult: FetchResult?

    private let wrapper = PurchaselyWrapper.shared

    private let userIdKey = "user_id"
    private let themeKey = "theme_mode"
    private let displayModeKey = "display_mode"
    private let consentAnalyticsKey = "consent_analytics"
    private let consentIdentifiedAnalyticsKey = "consent_identified_analytics"
    private let consentPersonalizationKey = "consent_personalization"
    private let consentCampaignsKey = "consent_campaigns"
    private let consentThirdPartyKey = "consent_third_party"

    var sdkVersion: String { wrapper.sdkVersion }

    init() {
        userId = UserDefaults.standard.string(forKey: userIdKey)
        themeMode = UserDefaults.standard.string(forKey: themeKey) ?? "system"
        sdkMode = PurchaselySDKMode.current()
        sdkModeRestartMessage = nil

        let defaults = UserDefaults.standard
        analyticsConsent = defaults.object(forKey: consentAnalyticsKey) == nil ? true : defaults.bool(forKey: consentAnalyticsKey)
        identifiedAnalyticsConsent = defaults.object(forKey: consentIdentifiedAnalyticsKey) == nil ? true : defaults.bool(forKey: consentIdentifiedAnalyticsKey)
        personalizationConsent = defaults.object(forKey: consentPersonalizationKey) == nil ? true : defaults.bool(forKey: consentPersonalizationKey)
        campaignsConsent = defaults.object(forKey: consentCampaignsKey) == nil ? true : defaults.bool(forKey: consentCampaignsKey)
        thirdPartyConsent = defaults.object(forKey: consentThirdPartyKey) == nil ? true : defaults.bool(forKey: consentThirdPartyKey)
        runningMode = RunningModeRepository.shared.isObserverMode ? "observer" : "full"
        displayMode = UserDefaults.standard.string(forKey: displayModeKey) ?? "fullscreen"

        applyConsentPreferences()
        anonymousId = wrapper.anonymousUserId
    }

    func refreshAnonymousId() {
        anonymousId = wrapper.anonymousUserId
    }

    func prefetchOnboardingPresentation() {
        Task {
            onboardingFetchResult = await wrapper.loadPresentation(placementId: "onboarding") { result in
                if case .purchased = result { PremiumManager.shared.refreshPremiumStatus() }
                if case .restored = result { PremiumManager.shared.refreshPremiumStatus() }
            }
        }
    }

    func displayOnboardingPaywall(from viewController: UIViewController?) {
        guard case .success(let presentation) = onboardingFetchResult else { return }
        wrapper.display(presentation: presentation, from: viewController)
    }

    func login(userId: String) {
        guard !userId.isEmpty else { return }

        wrapper.userLogin(userId: userId) { refresh in
            if refresh {
                PremiumManager.shared.refreshPremiumStatus()
            }
            print("[Shaker] Logged in as: \(userId) (refresh: \(refresh))")
        }

        self.userId = userId
        UserDefaults.standard.set(userId, forKey: userIdKey)
        wrapper.setUserAttribute(userId, forKey: "user_id")
    }

    func logout() {
        wrapper.userLogout()
        userId = nil
        UserDefaults.standard.removeObject(forKey: userIdKey)
        PremiumManager.shared.refreshPremiumStatus()
        print("[Shaker] Logged out")
    }

    func restorePurchases() {
        restoreMessage = nil
        wrapper.restoreAllProducts(
            success: { [weak self] in
                PremiumManager.shared.refreshPremiumStatus()
                DispatchQueue.main.async {
                    self?.restoreMessage = "Purchases restored successfully!"
                }
                print("[Shaker] Restore success")
            },
            failure: { [weak self] error in
                DispatchQueue.main.async {
                    self?.restoreMessage = error.localizedDescription
                }
                print("[Shaker] Restore error: \(error.localizedDescription)")
            }
        )
    }

    func clearRestoreMessage() {
        restoreMessage = nil
    }

    func setThemeMode(_ mode: String) {
        themeMode = mode
        UserDefaults.standard.set(mode, forKey: themeKey)
        wrapper.setUserAttribute(mode, forKey: "app_theme")
    }

    func setSdkMode(_ mode: PurchaselySDKMode) {
        guard sdkMode != mode else { return }

        sdkMode = mode
        mode.persist()
        NotificationCenter.default.post(name: .purchaselySdkModeDidChange, object: nil)
        sdkModeRestartMessage =
            "Purchasely SDK switched to \(mode.title). Please kill and relaunch the app."
        print("[Shaker] SDK mode updated to \(mode.rawValue)")
    }

    func clearSdkModeRestartMessage() {
        sdkModeRestartMessage = nil
    }

    func setAnalyticsConsent(_ enabled: Bool) {
        analyticsConsent = enabled
        UserDefaults.standard.set(enabled, forKey: consentAnalyticsKey)
        applyConsentPreferences()
    }

    func setIdentifiedAnalyticsConsent(_ enabled: Bool) {
        identifiedAnalyticsConsent = enabled
        UserDefaults.standard.set(enabled, forKey: consentIdentifiedAnalyticsKey)
        applyConsentPreferences()
    }

    func setPersonalizationConsent(_ enabled: Bool) {
        personalizationConsent = enabled
        UserDefaults.standard.set(enabled, forKey: consentPersonalizationKey)
        applyConsentPreferences()
    }

    func setCampaignsConsent(_ enabled: Bool) {
        campaignsConsent = enabled
        UserDefaults.standard.set(enabled, forKey: consentCampaignsKey)
        applyConsentPreferences()
    }

    func setThirdPartyConsent(_ enabled: Bool) {
        thirdPartyConsent = enabled
        UserDefaults.standard.set(enabled, forKey: consentThirdPartyKey)
        applyConsentPreferences()
    }

    func setRunningMode(_ mode: String) {
        runningMode = mode
        RunningModeRepository.shared.runningMode = mode == "observer" ? .paywallObserver : .full
        print("[Shaker] Running mode changed to: \(mode)")
    }

    func setDisplayMode(_ mode: String) {
        displayMode = mode
        UserDefaults.standard.set(mode, forKey: displayModeKey)
        print("[Shaker] Display mode changed to: \(mode)")
    }

    private func applyConsentPreferences() {
        var revoked = Set<PLYDataProcessingPurpose>()
        if !analyticsConsent { revoked.insert(.analytics) }
        if !identifiedAnalyticsConsent { revoked.insert(.identifiedAnalytics) }
        if !personalizationConsent { revoked.insert(.personalization) }
        if !campaignsConsent { revoked.insert(.campaigns) }
        if !thirdPartyConsent { revoked.insert(.thirdPartyIntegrations) }
        wrapper.revokeDataProcessingConsent(for: revoked)
        print("[Shaker] Consent updated — revoked: \(revoked)")
    }
}
